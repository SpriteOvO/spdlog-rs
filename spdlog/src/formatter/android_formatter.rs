// TODO: Remove this file, use `PatternFormatter` instead
//
// Need to keep waiting for conditional space and brackets to be supported in
// pattern template strings (optional fields require these, e.g. `logger_name`)

use std::fmt::{self, Write};

use crate::{
    formatter::{Formatter, FormatterContext},
    Error, Record, StringBuf,
};

#[derive(Clone)]
pub(crate) struct AndroidFormatter {}

impl AndroidFormatter {
    #[must_use]
    pub(crate) fn new() -> Self {
        Self {}
    }

    fn format_impl(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        _ctx: &mut FormatterContext,
    ) -> Result<(), fmt::Error> {
        #[cfg(not(feature = "flexible-string"))]
        dest.reserve(crate::string_buf::RESERVE_SIZE);

        if let Some(logger_name) = record.logger_name() {
            dest.write_str("[")?;
            dest.write_str(logger_name)?;
            dest.write_str("] ")?;
        }

        if let Some(srcloc) = record.source_location() {
            dest.write_str("[")?;
            dest.write_str(srcloc.module_path())?;
            dest.write_str(", ")?;
            dest.write_str(srcloc.file())?;
            dest.write_str(":")?;
            dest.write_str(&numtoa::BaseN::<10>::u32(srcloc.line()))?;
            dest.write_str("] ")?;
        }

        dest.write_str(record.payload())?;

        record.key_values().write_to(dest, true)?;
        Ok(())
    }
}

impl Formatter for AndroidFormatter {
    fn format(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        ctx: &mut FormatterContext,
    ) -> crate::Result<()> {
        self.format_impl(record, dest, ctx)
            .map_err(Error::FormatRecord)
    }
}

impl Default for AndroidFormatter {
    fn default() -> Self {
        Self::new()
    }
}
