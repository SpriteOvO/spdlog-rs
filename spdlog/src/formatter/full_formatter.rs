//! Provides a full info formatter.

use std::fmt::{self, Write};

use cfg_if::cfg_if;

use crate::{
    formatter::{FmtExtraInfo, Formatter, LOCAL_TIME_CACHER},
    Error, Record, StringBuf, __EOL,
};

#[rustfmt::skip]
/// Full information logs formatter.
///
/// It is the default formatter for most sinks.
///
/// Log messages formatted by it look like:
///
///  - Default:
///
///    <pre>
///    [2022-11-02 09:23:12.263] [<font color="#0DBC79">info</font>] hello, world!
///    </pre>
///
///  - If the logger has a name:
///
///    <pre>
///    [2022-11-02 09:23:12.263] [logger-name] [<font color="#0DBC79">info</font>] hello, world!
///    </pre>
/// 
///  - If crate feature `source-location` is enabled:
///
///    <pre>
///    [2022-11-02 09:23:12.263] [logger-name] [<font color="#0DBC79">info</font>] [mod::path, src/main.rs:4] hello, world!
///    </pre>
#[derive(Clone)]
pub struct FullFormatter {
    with_eol: bool,
}

impl FullFormatter {
    /// Constructs a `FullFormatter`.
    #[must_use]
    pub fn new() -> FullFormatter {
        FullFormatter { with_eol: true }
    }

    #[must_use]
    pub(crate) fn without_eol() -> Self {
        Self { with_eol: false }
    }

    fn format_impl(
        &self,
        record: &Record,
        dest: &mut StringBuf,
    ) -> Result<FmtExtraInfo, fmt::Error> {
        cfg_if! {
            if #[cfg(not(feature = "flexible-string"))] {
                dest.reserve(crate::string_buf::RESERVE_SIZE);
            }
        }

        {
            let mut local_time_cacher = LOCAL_TIME_CACHER.lock();
            let time = local_time_cacher.get(record.time());
            dest.write_str("[")?;
            dest.write_str(&time.full_second_str())?;
            dest.write_str(".")?;
            write!(dest, "{:03}", time.millisecond())?;
            dest.write_str("] [")?;
        }

        if let Some(logger_name) = record.logger_name() {
            dest.write_str(logger_name)?;
            dest.write_str("] [")?;
        }

        let style_range_begin = dest.len();

        dest.write_str(record.level().as_str())?;

        let style_range_end = dest.len();

        if let Some(srcloc) = record.source_location() {
            dest.write_str("] [")?;
            dest.write_str(srcloc.module_path())?;
            dest.write_str(", ")?;
            dest.write_str(srcloc.file())?;
            dest.write_str(":")?;
            write!(dest, "{}", srcloc.line())?;
        }

        dest.write_str("] ")?;
        dest.write_str(record.payload())?;

        if self.with_eol {
            dest.write_str(__EOL)?;
        }

        Ok(FmtExtraInfo {
            style_range: Some(style_range_begin..style_range_end),
        })
    }
}

impl Formatter for FullFormatter {
    fn format(&self, record: &Record, dest: &mut StringBuf) -> crate::Result<FmtExtraInfo> {
        self.format_impl(record, dest).map_err(Error::FormatRecord)
    }

    fn clone_box(&self) -> Box<dyn Formatter> {
        Box::new(self.clone())
    }
}

impl Default for FullFormatter {
    fn default() -> FullFormatter {
        FullFormatter::new()
    }
}

#[cfg(test)]
mod tests {
    use chrono::prelude::*;

    use super::*;
    use crate::{Level, __EOL};

    #[test]
    fn format() {
        let record = Record::new(Level::Warn, "test log content");
        let mut buf = StringBuf::new();
        let extra_info = FullFormatter::new().format(&record, &mut buf).unwrap();

        let local_time: DateTime<Local> = record.time().into();
        assert_eq!(
            format!(
                "[{}] [warn] test log content{}",
                local_time.format("%Y-%m-%d %H:%M:%S.%3f"),
                __EOL
            ),
            buf
        );
        assert_eq!(Some(27..31), extra_info.style_range());
    }
}
