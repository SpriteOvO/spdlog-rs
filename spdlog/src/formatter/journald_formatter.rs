// TODO: Remove this file, use `PatternFormatter` instead
//
// Need to keep waiting for conditional space and brackets to be supported in
// pattern template strings (optional fields require these, e.g. `logger_name`)

use std::fmt::{self, Write};

use cfg_if::cfg_if;

use crate::{
    formatter::{Formatter, FormatterContext},
    Error, Record, StringBuf, __EOL,
};

#[derive(Clone)]
pub(crate) struct JournaldFormatter {}

impl JournaldFormatter {
    #[must_use]
    pub(crate) fn new() -> Self {
        Self {}
    }

    fn format_impl(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        ctx: &mut FormatterContext,
    ) -> Result<(), fmt::Error> {
        cfg_if! {
            if #[cfg(not(feature = "flexible-string"))] {
                dest.reserve(crate::string_buf::RESERVE_SIZE);
            }
        }

        dest.write_str("[")?;

        if let Some(logger_name) = record.logger_name() {
            dest.write_str(logger_name)?;
            dest.write_str("] [")?;
        }

        let style_range_begin = dest.len();

        dest.write_str(record.level().as_str())?;

        let style_range_end = dest.len();

        dest.write_str("] ")?;
        dest.write_str(record.payload())?;

        let kvs = record.key_values();
        if !kvs.is_empty() {
            dest.write_str(" { ")?;

            let mut iter = kvs.peekable();
            while let Some((key, value)) = iter.next() {
                dest.write_str(key.as_str())?;
                dest.write_str("=")?;
                write!(dest, "{}", value)?;
                if iter.peek().is_some() {
                    dest.write_str(", ")?;
                }
            }
            dest.write_str(" }")?;
        }

        dest.write_str(__EOL)?;

        ctx.set_style_range(Some(style_range_begin..style_range_end));
        Ok(())
    }
}

impl Formatter for JournaldFormatter {
    fn format(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        ctx: &mut FormatterContext,
    ) -> crate::Result<()> {
        self.format_impl(record, dest, ctx)
            .map_err(Error::FormatRecord)
    }
}

impl Default for JournaldFormatter {
    fn default() -> Self {
        Self::new()
    }
}
