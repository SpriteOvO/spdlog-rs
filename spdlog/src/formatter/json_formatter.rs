use std::{
    fmt::{self, Write},
    marker::PhantomData,
    time::SystemTime,
};

use cfg_if::cfg_if;
use serde::{
    ser::{SerializeMap, SerializeStruct},
    Serialize, Serializer,
};

use crate::{
    formatter::{Formatter, FormatterContext},
    Error, Record, StringBuf, __EOL,
};

fn opt_to_num<T>(opt: Option<T>) -> usize {
    opt.map_or(0, |_| 1)
}

struct JsonRecord<'a, 'b>(&'a Record<'b>);

impl Serialize for JsonRecord<'_, '_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let fields_len =
            4 + opt_to_num(self.0.logger_name()) + opt_to_num(self.0.source_location());
        let mut record = serializer.serialize_struct("JsonRecord", fields_len)?;

        record.serialize_field("level", &self.0.level())?;
        record.serialize_field(
            "timestamp",
            &self
                .0
                .time()
                .duration_since(SystemTime::UNIX_EPOCH)
                .ok()
                // https://github.com/SpriteOvO/spdlog-rs/pull/69#discussion_r1694063293
                .and_then(|dur| u64::try_from(dur.as_millis()).ok())
                .expect("invalid timestamp"),
        )?;
        record.serialize_field("payload", self.0.payload())?;

        if !self.0.key_values().is_empty() {
            struct JsonKV<'a, 'b>(&'a Record<'b>);

            impl Serialize for JsonKV<'_, '_> {
                fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                where
                    S: Serializer,
                {
                    let kv = self.0.key_values();
                    let mut map = serializer.serialize_map(Some(kv.len()))?;
                    for (key, value) in kv {
                        map.serialize_entry(key.as_str(), &value)?;
                    }
                    map.end()
                }
            }

            record.serialize_field("kv", &JsonKV(self.0))?;
        }

        if let Some(logger_name) = self.0.logger_name() {
            record.serialize_field("logger", logger_name)?;
        }
        record.serialize_field("tid", &self.0.tid())?;
        if let Some(src_loc) = self.0.source_location() {
            record.serialize_field("source", src_loc)?;
        }

        record.end()
    }
}

enum JsonFormatterError {
    Fmt(fmt::Error),
    Serialization(serde_json::Error),
}

impl From<fmt::Error> for JsonFormatterError {
    fn from(value: fmt::Error) -> Self {
        JsonFormatterError::Fmt(value)
    }
}

impl From<serde_json::Error> for JsonFormatterError {
    fn from(value: serde_json::Error) -> Self {
        JsonFormatterError::Serialization(value)
    }
}

impl From<JsonFormatterError> for crate::Error {
    fn from(value: JsonFormatterError) -> Self {
        match value {
            JsonFormatterError::Fmt(e) => Error::FormatRecord(e),
            JsonFormatterError::Serialization(e) => Error::SerializeRecord(e.into()),
        }
    }
}

#[rustfmt::skip]
/// JSON logs formatter.
/// 
/// Each log will be serialized into a single line of JSON object with the following schema.
/// 
/// ## Schema
/// 
/// | Field       | Type         | Description                                                                                                                    |
/// |-------------|--------------|--------------------------------------------------------------------------------------------------------------------------------|
/// | `level`     | String       | The level of the log. Same as the return of [`Level::as_str`].                                                                 |
/// | `timestamp` | Integer(u64) | The timestamp when the log was generated, in milliseconds since January 1, 1970 00:00:00 UTC.                                  |
/// | `payload`   | String       | The contents of the log.                                                                                                       |
/// | `kv`        | Object/Null  | The key-values of the log. Null if kv is not specified.                                                                        |
/// | `logger`    | String/Null  | The name of the logger. Null if the logger has no name.                                                                        |
/// | `tid`       | Integer(u64) | The thread ID when the log was generated.                                                                                      |
/// | `source`    | Object/Null  | The source location of the log. See [`SourceLocation`] for its schema. Null if crate feature `source-location` is not enabled. |
/// 
/// <div class="warning">
/// 
/// - If the type of a field is Null, the field will not be present or be `null`.
/// 
/// - The order of the fields is not guaranteed.
/// 
/// </div>
/// 
/// ---
/// 
/// ## Examples
///
///  - Default:
/// 
///    ```json
///    {"level":"info","timestamp":1722817424798,"payload":"hello, world!","tid":3472525}
///    {"level":"error","timestamp":1722817424798,"payload":"something went wrong","tid":3472525}
///    ```
/// 
///  - If the logger has a name:
/// 
///    ```json
///    {"level":"info","timestamp":1722817541459,"payload":"hello, world!","logger":"app-component","tid":3478045}
///    {"level":"error","timestamp":1722817541459,"payload":"something went wrong","logger":"app-component","tid":3478045}
///    ```
/// 
///  - If key-values are present:
/// 
///    ```json
///    {"level":"info","timestamp":1722817541459,"payload":"hello, world!","kv":{"k1":123,"k2":"cool"},"tid":3478045}
///    {"level":"error","timestamp":1722817541459,"payload":"something went wrong","kv":{"k1":123,"k2":"cool"},"tid":3478045}
///    ```
/// 
///  - If crate feature `source-location` is enabled:
/// 
///    ```json
///    {"level":"info","timestamp":1722817572709,"payload":"hello, world!","tid":3479856,"source":{"module_path":"my_app::say_hi","file":"src/say_hi.rs","line":4,"column":5}}
///    {"level":"error","timestamp":1722817572709,"payload":"something went wrong","tid":3479856,"source":{"module_path":"my_app::say_hi","file":"src/say_hi.rs","line":5,"column":5}}
///    ```
/// 
/// [`Level::as_str`]: crate::Level::as_str
/// [`SourceLocation`]: crate::SourceLocation
#[derive(Clone)]
pub struct JsonFormatter(PhantomData<()>);

impl JsonFormatter {
    /// Constructs a `JsonFormatter`.
    #[must_use]
    pub fn new() -> JsonFormatter {
        JsonFormatter(PhantomData)
    }

    fn format_impl(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        _ctx: &mut FormatterContext,
    ) -> Result<(), JsonFormatterError> {
        cfg_if! {
            if #[cfg(not(feature = "flexible-string"))] {
                dest.reserve(crate::string_buf::RESERVE_SIZE);
            }
        }

        // TODO: https://github.com/serde-rs/json/issues/863
        //
        // The performance can be significantly optimized here if the issue can be
        // solved.
        dest.write_str(&serde_json::to_string(&JsonRecord(record))?)?;

        dest.write_str(__EOL)?;

        Ok(())
    }
}

impl Formatter for JsonFormatter {
    fn format(
        &self,
        record: &Record,
        dest: &mut StringBuf,
        ctx: &mut FormatterContext,
    ) -> crate::Result<()> {
        self.format_impl(record, dest, ctx).map_err(Into::into)
    }
}

impl Default for JsonFormatter {
    fn default() -> Self {
        JsonFormatter::new()
    }
}

#[cfg(test)]
mod tests {
    use chrono::prelude::*;

    use super::*;
    use crate::{kv, Level, SourceLocation, __EOL};

    #[test]
    fn should_format_json() {
        let mut dest = StringBuf::new();
        let formatter = JsonFormatter::new();
        let record = Record::new(Level::Info, "payload", None, None, &[]);
        let mut ctx = FormatterContext::new();
        formatter.format(&record, &mut dest, &mut ctx).unwrap();

        let local_time: DateTime<Local> = record.time().into();

        assert_eq!(ctx.style_range(), None);
        assert_eq!(
            dest.to_string(),
            format!(
                r#"{{"level":"info","timestamp":{},"payload":"{}","tid":{}}}{}"#,
                local_time.timestamp_millis(),
                "payload",
                record.tid(),
                __EOL
            )
        );
    }

    #[test]
    fn should_format_json_with_logger_name() {
        let mut dest = StringBuf::new();
        let formatter = JsonFormatter::new();
        let record = Record::new(Level::Info, "payload", None, Some("my-component"), &[]);
        let mut ctx = FormatterContext::new();
        formatter.format(&record, &mut dest, &mut ctx).unwrap();

        let local_time: DateTime<Local> = record.time().into();

        assert_eq!(ctx.style_range(), None);
        assert_eq!(
            dest.to_string(),
            format!(
                r#"{{"level":"info","timestamp":{},"payload":"{}","logger":"my-component","tid":{}}}{}"#,
                local_time.timestamp_millis(),
                "payload",
                record.tid(),
                __EOL
            )
        );
    }

    #[test]
    fn should_format_json_with_src_loc() {
        let mut dest = StringBuf::new();
        let formatter = JsonFormatter::new();
        let record = Record::new(
            Level::Info,
            "payload",
            Some(SourceLocation::__new("module", "file.rs", 1, 2)),
            None,
            &[],
        );
        let mut ctx = FormatterContext::new();
        formatter.format(&record, &mut dest, &mut ctx).unwrap();

        let local_time: DateTime<Local> = record.time().into();

        assert_eq!(ctx.style_range(), None);
        assert_eq!(
            dest.to_string(),
            format!(
                r#"{{"level":"info","timestamp":{},"payload":"{}","tid":{},"source":{{"module_path":"module","file":"file.rs","line":1,"column":2}}}}{}"#,
                local_time.timestamp_millis(),
                "payload",
                record.tid(),
                __EOL
            )
        );
    }

    #[test]
    fn should_format_json_with_kv() {
        let mut dest = StringBuf::new();
        let formatter = JsonFormatter::new();
        let kvs = [
            (kv::Key::__from_static_str("k1"), kv::Value::from(114)),
            (kv::Key::__from_static_str("k2"), kv::Value::from("514")),
        ];
        let record = Record::new(Level::Info, "payload", None, None, &kvs);
        let mut ctx = FormatterContext::new();
        formatter.format(&record, &mut dest, &mut ctx).unwrap();

        let local_time: DateTime<Local> = record.time().into();

        assert_eq!(ctx.style_range(), None);
        assert_eq!(
            dest.to_string(),
            format!(
                r#"{{"level":"info","timestamp":{},"payload":"{}","kv":{{"k1":114,"k2":"514"}},"tid":{}}}{}"#,
                local_time.timestamp_millis(),
                "payload",
                record.tid(),
                __EOL
            )
        );
    }
}
