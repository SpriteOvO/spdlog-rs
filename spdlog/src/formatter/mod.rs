//! Provides formatters for sink formatting log records.
//!
//! # Formatter
//!
//! Each normal *Sink* owns a *Formatter*, which is used to format each log.
//!
//! The default formatter for most sinks is [`FullFormatter`], you can call
//! [`Sink::set_formatter`] to replace it with another formatter.
//!
//! The easiest way to make a custom formatter is to build a pattern, see
//! [Compile-time and runtime pattern
//! formatter](#compile-time-and-runtime-pattern-formatter) below. If pattern
//! isn't flexible enough for you, you need to implement [`Formatter`] trait for
//! your own formatter struct. See the implementation of [`FullFormatter`] and
//! [./examples] directory for examples.
//!
//! # Compile-time and runtime pattern formatter
//!
//! *spdlog-rs* supports formatting your log records according to a pattern
//! string. There are 2 ways to construct a pattern:
//!
//! - Macro [`pattern!`]: Builds a pattern at compile-time.
//! - Macro [`runtime_pattern!`]: Builds a pattern at runtime.
//!
//! ```
//! use spdlog::formatter::{pattern, PatternFormatter};
//! # use spdlog::sink::{Sink, WriteSink};
//!
//! # fn main() -> Result<(), Box<dyn std::error::Error>> {
//! // This pattern is built at compile-time, the template accepts only a literal string.
//! let pattern = pattern!("[{date} {time}.{millisecond}] [{level}] {payload}{eol}");
//!
//! #[cfg(feature = "runtime-pattern")]
//! {
//!     use spdlog::formatter::runtime_pattern;
//!
//!     // This pattern is built at runtime, the template accepts a runtime string.
//!     let input = "[{date} {time}.{millisecond}] [{level}] {payload}{eol}";
//!     let pattern = runtime_pattern!(input)?;
//! }
//!
//! // Use the compile-time or runtime pattern.
//! # let your_sink = WriteSink::builder().target(vec![]).build()?;
//! your_sink.set_formatter(Box::new(PatternFormatter::new(pattern)));
//! # Ok(()) }
//! ```
//!
//! [`Sink::set_formatter`]: crate::sink::Sink::set_formatter
//! [./examples]: https://github.com/SpriteOvO/spdlog-rs/tree/main/spdlog/examples

mod full_formatter;
#[cfg(any(
    all(target_os = "linux", feature = "native", feature = "libsystemd"),
    all(doc, not(doctest))
))]
mod journald_formatter;
mod local_time_cacher;
mod pattern_formatter;

use std::ops::Range;

pub use full_formatter::*;
#[cfg(any(
    all(target_os = "linux", feature = "native", feature = "libsystemd"),
    all(doc, not(doctest))
))]
pub(crate) use journald_formatter::*;
pub(crate) use local_time_cacher::*;
pub use pattern_formatter::*;

use crate::{Record, Result, StringBuf};

/// Represents a formatter that can be used for formatting logs.
///
/// # Examples
///
/// See the implementation of [`FullFormatter`] and [./examples] directory.
///
/// [./examples]: https://github.com/SpriteOvO/spdlog-rs/tree/main/spdlog/examples
pub trait Formatter: Send + Sync {
    /// Formats a log record.
    fn format(&self, record: &Record, dest: &mut StringBuf) -> Result<FmtExtraInfo>;

    /// Clones self into a boxed trait object.
    #[must_use]
    fn clone_box(&self) -> Box<dyn Formatter>;
}

/// Extra information for formatted text.
#[derive(Clone, Eq, PartialEq, Hash, Debug, Default)]
pub struct FmtExtraInfo {
    style_range: Option<Range<usize>>,
}

impl FmtExtraInfo {
    /// Constructs a `FmtExtraInfo`.
    #[must_use]
    pub fn new() -> FmtExtraInfo {
        FmtExtraInfo::default()
    }

    /// Gets a [`FmtExtraInfoBuilder`].
    #[must_use]
    pub fn builder() -> FmtExtraInfoBuilder {
        FmtExtraInfoBuilder::new()
    }

    /// A style range (in bytes) of the formatted text.
    ///
    /// If style is available in the sink, the text in the range will be
    /// rendered in the style corresponding to that log message level, otherwise
    /// it will be ignored.
    ///
    /// Its indexes are guaranteed by the setter to be the correct UTF-8
    /// boundary.
    #[must_use]
    pub fn style_range(&self) -> Option<Range<usize>> {
        self.style_range.clone() // This clone is cheap
    }
}

#[allow(missing_docs)]
#[derive(Clone, Eq, PartialEq, Hash, Debug, Default)]
pub struct FmtExtraInfoBuilder {
    info: FmtExtraInfo,
}

impl FmtExtraInfoBuilder {
    /// Constructs a `FmtExtraInfoBuilder`.
    ///
    /// The default value of [`FmtExtraInfo`] is the same as
    /// [`FmtExtraInfo::new`].
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Sets style range (in bytes) of the formatted text.
    ///
    /// Users must ensure that indexes are correctly UTF-8 boundary.
    #[must_use]
    pub fn style_range(mut self, range: Range<usize>) -> Self {
        self.info.style_range = Some(range);
        self
    }

    /// Builds a [`FmtExtraInfo`].
    #[must_use]
    pub fn build(self) -> FmtExtraInfo {
        self.info
    }
}
