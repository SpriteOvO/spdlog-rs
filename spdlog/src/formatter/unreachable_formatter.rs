use std::marker::PhantomData;

use crate::{
    formatter::{Formatter, FormatterContext},
    Record, StringBuf,
};

#[derive(Clone)]
pub(crate) struct UnreachableFormatter(PhantomData<()>);

impl UnreachableFormatter {
    pub(crate) fn new() -> Self {
        Self(PhantomData)
    }
}

impl Formatter for UnreachableFormatter {
    fn format(
        &self,
        _record: &Record,
        _dest: &mut StringBuf,
        _ctx: &mut FormatterContext,
    ) -> crate::Result<()> {
        unreachable!()
    }
}
