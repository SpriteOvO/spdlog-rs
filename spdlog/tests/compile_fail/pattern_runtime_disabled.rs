use spdlog::formatter::runtime_pattern;

fn runtime_pattern() {
    runtime_pattern!("{logger}");
}

fn main() {}
