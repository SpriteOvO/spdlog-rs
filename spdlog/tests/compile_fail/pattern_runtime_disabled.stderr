error: proc macro panicked
 --> tests/compile_fail/pattern_runtime_disabled.rs:4:5
  |
4 |     runtime_pattern!("{logger}");
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: macro `runtime_pattern` required to enable crate feature `runtime-pattern` for spdlog-rs
